[Instrumentation.Http] Added an example for metrics to the document. (#4216)
open-telemetry/opentelemetry-dotnet: src/OpenTelemetry.Instrumentation.Http/README.md

HTTP instrumentation must be enabled at application startup.

> #### For Trace

Reviewer: ```suggestion
#### Traces
```

> The following example demonstrates adding HTTP client instrumentation with the

Reviewer 2: ```suggestion
The following example demonstrates adding `HttpClient` instrumentation with the
```

> extension method `.AddHttpClientInstrumentation()` to a console application.

Reviewer: ```suggestion
extension method `.AddHttpClientInstrumentation()` on `TracerProviderBuilder` to a console application.
```

}
```

> #### Metrics

Reviewer: please put a note that metrics are not available for .NET Framework.

> The following example demonstrates adding HTTP client instrumentation with the

Reviewer 2: ```suggestion
The following example demonstrates adding `HttpClient` instrumentation with the
```

extension method `.AddHttpClientInstrumentation()` on `MeterProviderBuilder` to
a console application. This example also sets up the OpenTelemetry Console
Exporter, which requires adding the package
[`OpenTelemetry.Exporter.Console`](../OpenTelemetry.Exporter.Console/README.md)
to the application.

```csharp
using OpenTelemetry;
using OpenTelemetry.Metrics;

> public class Program

Reviewer 3: Instead of using the explicit class and Main entry point, we might want to shift to a simpler model https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started-aspnetcore/Program.cs. Doesn't block this PR though.

{
    public static void Main(string[] args)
    {
        using var meterProvider = Sdk.CreateMeterProviderBuilder()
            .AddHttpClientInstrumentation()
            .AddConsoleExporter()
            .Build();
    }
}
```

> Check how to enable `OpenTelemetry.Instrumentation.Http` for Trace

Reviewer: and similar for asp.net

> Check how to enable `OpenTelemetry.Instrumentation.Http` for Trace

Author: updated

and Metrics with this [example](../../examples/AspNetCore/Program.cs).

For an ASP.NET application, adding instrumentation is typically done in the
`Global.asax.cs`. Refer to the documentation for
[OpenTelemetry.Instrumentation.AspNet](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/README.md).

> ### List of Metrics produced

Reviewer 3: minor:

```suggestion
### List of metrics produced
```

> ### List of metrics produced

Reviewer: why "###" instead of ## or put this as #### under metrics.

}
```

#### Metrics

The following example demonstrates adding HTTP client instrumentation with the
extension method `.AddHttpClientInstrumentation()` on `MeterProviderBuilder` to
a console application. This example also sets up the OpenTelemetry Console
Exporter, which requires adding the package
[`OpenTelemetry.Exporter.Console`](../OpenTelemetry.Exporter.Console/README.md)
to the application.

```csharp
using OpenTelemetry;
using OpenTelemetry.Metrics;

public class Program
{
    public static void Main(string[] args)
    {
        using var meterProvider = Sdk.CreateMeterProviderBuilder()
            .AddHttpClientInstrumentation()
            .AddConsoleExporter()
            .Build();
    }
}
```

For an ASP.NET Core application, adding instrumentation is typically done in the
`ConfigureServices` of your `Startup` class. Refer to documentation for
[OpenTelemetry.Instrumentation.AspNetCore](../OpenTelemetry.Instrumentation.AspNetCore/README.md).

> In addition, check how to enable `OpenTelemetry.Instrumentation.Http` for Trace

Reviewer: I'd suggest to remove the previous 2 sentences. Its not really needed here, since we are linking to a proper example app.

> and Metrics with this [demo](../../examples/AspNetCore/Program.cs).

Reviewer: ```suggestion
and Metrics with this [example](../../examples/AspNetCore/Program.cs).
```

}
```

> #### For Metrics

Reviewer: ```suggestion
#### Metrics
```

The following example demonstrates adding HTTP client instrumentation with the
extension method `.AddHttpClientInstrumentation()` on `MeterProviderBuilder` to
a console application.

```csharp
using OpenTelemetry;
using OpenTelemetry.Metrics;

public class Program
{
    public static void Main(string[] args)
    {
        using var meterProvider = Sdk.CreateMeterProviderBuilder()
            .AddHttpClientInstrumentation()
            .AddConsoleExporter()
            .Build();
    }
}
```

> ### Metrics

Reviewer: ```suggestion
### List of Metrics produced
```

The instrumentation was implemented based on [metrics semantic
conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpclientduration).
Currently, the instrumentation supports the following metric.

| Name  | Instrument Type | Unit | Description | Attributes |
|-------|-----------------|------|-------------|------------|
| `http.client.duration` | Histogram | `ms` | Measures the duration of outbound HTTP requests. | http.method, http.scheme, http.status_code, http.flavor, net.peer.name |

> ### .NET Core & .NET

Reviewer: this heading seems misplaced.?

Author: fixed

For an ASP.NET Core application, adding instrumentation is typically done in the
`ConfigureServices` of your `Startup` class. Refer to documentation for
[OpenTelemetry.Instrumentation.AspNetCore](../OpenTelemetry.Instrumentation.AspNetCore/README.md).
In addition, check how to enable `OpenTelemetry.Instrumentation.Http` for Trace
and Metrics with this [demo](../../examples/AspNetCore/Program.cs).

For an ASP.NET application, adding instrumentation is typically done in the
`Global.asax.cs`. Refer to the documentation for
[OpenTelemetry.Instrumentation.AspNet](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/README.md).

> ## Advanced configuration

Reviewer: we need this.

Author: removed accidentally; fixed

}
```

#### For Metrics

The following example demonstrates adding HTTP client instrumentation with the
extension method `.AddHttpClientInstrumentation()` to a console application.

> The metric and its attributes being implemented are following Opentelemetry

Reviewer: please move this outside this subheading, to keep this section focused only on the enablement steps.

> The metric and its attributes being implemented are following Opentelemetry

Reviewer 2: Could be rephrased something like "The instrumentation currently supports following metrics:"  (Followed by the table)

That way we can keep adding more in future if needed.

[metrics semantic
conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpclientduration).

| Name  | Instrument Type | Unit | Description | Attributes |
|-------|-----------------|------|-------------|------------|

> | `http.client.duration` | Histogram | `ms` | Measures the duration of outbound HTTP requests. | http.method, http.scheme, http.status_code, http.flavor, net.peer.name |

Reviewer 2: We don't have to add attributes here.

> | `http.client.duration` | Histogram | `ms` | Measures the duration of outbound HTTP requests. | http.method, http.scheme, http.status_code, http.flavor, net.peer.name |

Reviewer 2: Can add a link to metric here inside the table https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpclientduration

Author: Along with this comment: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4216#discussion_r1115147196
Do you mean that we can instead update the Attributes column in this table to be a link to the spec?

I am thinking maybe documenting what's currently being implemented in the package may be a helpful info for the users.
As some of the attributes are optional and we don't currently implement all of the attributes listed in the spec.

Reviewer 2: I meant to remove the attributes column entirely. And update the `Name` column to [http.client.duration](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpclientduration)

> I am thinking maybe documenting what's currently being implemented in the package may be a helpful info for the users.
As some of the attributes are optional and we don't currently implement all of the attributes listed in the spec.

That is a good point, we have the same thing on traces side. I think we should be consistent, we can revisit it adding it later when the conventions are stable. Would like to hear feedback from others as well..

Author: removed the attributes column for now for consistency
